42.md

```diff
@@ -56,11 +56,23 @@ Relays MUST exclude `kind: 22242` events from being broadcasted to any client.
 
 ## Protocol flow
 
-At any moment the relay may send an `AUTH` message to the client containing a challenge. After receiving that the client may decide to
-authenticate itself or not. The challenge is expected to be valid for the duration of the connection or until a next challenge is sent by
-the relay.
+At specific moments the relay may send an `AUTH` message to the client containing a challenge, in one of the following forms:
 
-The client may send an auth message right before performing an action for which it knows authentication will be required -- for example, right
+- `["AUTH", <challenge-string>]`: No third array element. It means relay requires authentication for any resource access.
+If applicable, it MUST be sent right after WebSocket `connection`.
+Client may have already sent a message before receiving this one, in that case relay MUST also send an `AUTH` message in one of the other forms below;
+- `["AUTH", <challenge-string>, "sub:<subscription_id>"]`: It references a previous client `REQ` message with a specific subscription id as third array element.
+It means client subscription requested atleast one event that requires prior authentication to be accessed.
```

Comment:
can we change this to "It means client subscription potentially requested at least one event"?

relays shouldn't be expected to run queries which could require auth. if clients are writing properly restricted filters, this should be easy for relays to immediately discern whether authentication is needed. if client's filters are too broad (ie don't specify kind) then that should be the client's problem, not the relay's problem.

Reply:
Good idea